### PR TITLE
Support custom scrollIntoView behaviour from user side.

### PR DESCRIPTION
**Pull request by ulion**
#### Is this adding or improving a _feature_ or fixing a _bug_?
Adding a feature.

#### What's the new behavior?
user code can custom the behaviour of scrollInfoView

e.g. now we can add following code to the Editable's prop:
```
scrollSelectionIntoView={(editor, domRange) => {
  const leafEl = domRange.startContainer.parentElement!
  leafEl.getBoundingClientRect = domRange.getBoundingClientRect.bind(domRange)
  scrollIntoView(leafEl, {
    scrollMode: 'if-needed',
    block: 'nearest', // block = start | center | end | nearest
  })
  delete leafEl.getBoundingClientRect
}}
```
Then this will work like if https://github.com/ianstormtaylor/slate/pull/3746 is merged, together with https://github.com/ianstormtaylor/slate/pull/4007 and the concern in https://github.com/ianstormtaylor/slate/pull/3678#issuecomment-754038215 without changing any existed user code's behaviour.
#### How does this change work?
By introduce a customizable prop `scrollSelectionIntoView` for Editable

#### Have you checked that...?

- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: This mainly supply a way user can fix or get what scrollIntoView effect without touch the slate-react source code.
Reviewers: @

### CODE REVIEW

File: packages/slate-react/src/components/editable.tsx [hunk 1]
```diff
@@ -110,6 +111,7 @@ export const Editable = (props: EditableProps) => {
     readOnly = false,
     renderElement,
     renderLeaf,
+    scrollSelectionIntoView,
```
Comment (ianstormtaylor) [inline packages/slate-react/src/components/editable.tsx:114]: @ulion thank you for this! Love this idea of making it customizable.

Can we make this prop have a default which implements the existing logic, ie:

```ts
scrollToSelection = defaultScrollToSelection,
```

Which is defined as:

```ts
const defaultScrollToSelection = (editor) => {
  const { selection } = editor
  const domRange = ReactEditor.toDOMRange(editor, selection)
  const leafEl = newDomRange.startContainer.parentElement!
  scrollIntoView(leafEl, {
    scrollMode: 'if-needed',
    boundary: el,
  })
}
```

File: packages/slate-react/src/components/editable.tsx [hunk 2]
```diff
@@ -199,11 +201,15 @@ export const Editable = (props: EditableProps) => {
...
-      })
+      if (scrollSelectionIntoView) {
+        scrollSelectionIntoView(editor, newDomRange)
+      } else {
+        const leafEl = newDomRange.startContainer.parentElement!
+        scrollIntoView(leafEl, {
+          scrollMode: 'if-needed',
+          boundary: el,
+        })
+      }
```
Comment (ianstormtaylor) [inline packages/slate-react/src/components/editable.tsx:212]: And then with that change above we could simplify this to just:

```ts
scrollToSelection(editor)
```

**Comment by ulion** [comment]
@ianstormtaylor modified as you suggested.

**Comment by chasefarmer2808** [comment]
What is the status of this MR?  I am needing this for a collaborative editor I am making with SlateJS and YJS to prevent the user's selection from scrolling into view when another collaborator performs an operation on the document.

**Comment by dylans** [comment]
> What is the status of this MR? I am needing this for a collaborative editor I am making with SlateJS and YJS to prevent the user's selection from scrolling into view when another collaborator performs an operation on the document.

@chasefarmer2808  It looks like it needs a rebase against the latest in main, and needs to address a small remaining part of the feedback. If you or @ulion want to clean up the PR I'll review and do my best to get it landed quickly.

**Comment by chasefarmer2808** [comment]
@ulion can you add me as a collaborator to your fork?

**Comment by ulion** [comment]
> @ulion can you add me as a collaborator to your fork?

done.

File: packages/slate-react/src/components/editable.tsx [hunk 3]
```diff
@@ -106,7 +107,7 @@ export const Editable = (props: EditableProps) => {
     readOnly = false,
     renderElement,
     renderLeaf,
-    renderPlaceholder = props => <DefaultPlaceholder {...props} />,
```
Comment (chasefarmer2808) [inline packages/slate-react/src/components/editable.tsx:109]: @ulion is there a reason for the removal of this prop?  It looks like it is used on line 1125.

**Review by chasefarmer2808 — COMMENTED** [review]
@ulion I keep on getting this error in editable.tsx when trying to make a commit:
`11:8  error  Unable to resolve path to module 'slate'  import/no-unresolved`
Any ideas how to resolve?

**Comment by chasefarmer2808** [comment]
Apologies, my author information was incorrect in my last commit and had to revert it.

**Comment by chasefarmer2808** [comment]
Ok, I've rebased and made some small changes.  Getting linting erros regarding `slate` imports.  Also getting build errors saying `Argument of type ReactEditor is not assignable to type Element`.  Not sure if this is due to the 0.6x updates.  Could use some assistance.

**Comment by ulion** [comment]
@chasefarmer2808 I just re-pushed the code which fits current main branch code. hope you don't mind, I saw there were some issues. Please double check current commit and make any necessary change to go.

**Comment by dylans** [comment]
Seems like this is failing on the same test that #3888 is failing on as well... will try to sort this out over the weekend.

**Review by dylans — APPROVED** [review]

**Comment by chasefarmer2808** [comment]
Thanks all!  I'm new to contributing to this project.  What happens now regarding getting this published to a release or to npm?

**Comment by dylans** [comment]
@chasefarmer2808 we pushed 0.65.3 today!

**Pull request merged**